src/backend/renderer/gles2/mod.rs

```diff
@@ -0,0 +1,568 @@
+use crate::backend::egl::{EGLContext, EGLSurface, ffi::egl::types::EGLImage};
+use super::{Renderer, Frame, Bind, Unbind, Transform, Texture};
+
+#[cfg(feature = "wayland_frontend")]
+use wayland_server::protocol::{wl_shm, wl_buffer};
+
+#[allow(clippy::all, missing_docs)]
+pub mod ffi {
+    include!(concat!(env!("OUT_DIR"), "/gl_bindings.rs"));
+}
+
+#[derive(Debug)]
+struct Gles2Program {
+    program: ffi::types::GLuint,
+    uniform_tex: ffi::types::GLint,
+    uniform_matrix: ffi::types::GLint,
+    uniform_invert_y: ffi::types::GLint,
+    uniform_alpha: ffi::types::GLint,
+    attrib_position: ffi::types::GLint,
+    attrib_tex_coords: ffi::types::GLint,
+}
+
+pub struct Gles2Texture {
+    texture: ffi::types::GLuint,
+    texture_kind: usize,
+    is_external: bool,
+    y_inverted: bool,
+    width: u32,
+    height: u32,
+}
+
+impl Texture for Gles2Texture {
+    fn width(&self) -> u32 {
+        self.width
+    }
+    fn height(&self) -> u32 {
+        self.height
+    }
+}
+
+#[derive(Clone)]
+struct WeakGles2Buffer {
+    dmabuf: WeakDmabuf,
+    image: EGLImage,
+    rbo: ffi::types::GLuint,
+    fbo: ffi::types::GLuint,
+}
+
+struct Gles2Buffer {
+    internal: WeakGles2Buffer,
+    _dmabuf: Dmabuf,
+}
+
+pub struct Gles2Renderer {
+    internal: Arc<Gles2RendererInternal>,
+    buffers: Vec<WeakGles2Buffer>,
+    current_buffer: Option<Gles2Buffer>,
+}
+
+struct Gles2RendererInternal {
+    gl: ffi::Gles2,
+    egl: EGLContext,
+    extensions: Vec<String>,
+    programs: [Gles2Program; shaders::FRAGMENT_COUNT],
+    logger: Option<*mut ::slog::Logger>,
+}
+
+pub struct Gles2Frame {
+    internal: Arc<Gles2RendererInternal>,
+    projection: Matrix3<f32>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Gles2Error {
+    #[error("Failed to compile Shader: {0}")]
+    ShaderCompileError(&'static str),
+    #[error("Failed to link Program")]
+    ProgramLinkError,
+    #[error("Failed to bind Framebuffer")]
+    FramebufferBindingError,
+    #[error("Failed to load GL functions from EGL")]
+    GLFunctionLoaderError,
+    /// The required GL extension is not supported by the underlying implementation
+    #[error("None of the following GL extensions is supported by the underlying GL implementation, at least one is required: {0:?}")]
+    GLExtensionNotSupported(&'static [&'static str]),
+    #[error("Failed to active egl context")]
+    ContextActivationError(#[from] crate::backend::egl::MakeCurrentError),
+    #[error("Failed to convert dmabuf to EGLImage")]
+    BindBufferEGLError(#[source] crate::backend::egl::Error),
+    #[error("Unsupported pixel format: {0:?}")]
+    #[cfg(feature = "wayland_frontend")]
+    UnsupportedPixelFormat(wl_shm::Format),
+    #[error("Error accessing the buffer ({0:?})")]
+    #[cfg(feature = "wayland_frontend")]
+    BufferAccessError(crate::wayland::shm::BufferAccessError),
+}
+
+extern "system" fn gl_debug_log(_source: ffi::types::GLenum,
```

Comment:
Why `extern "system"` rather than `extern "C"` ?

Reply:
Because that is what "gl_generator" generates and the api therefor expects (and to Rust "system" is different from "C", although this code will likely never run on anything else then linux anyway).